src/solvers/dgsem_p4est/dg_2d.jl

```diff
@@ -0,0 +1,420 @@
+            end
+            u_buffer[2:4, i] = u[2:4, i_large, j_large, element]
+
+            for v in eachvariable(equations)
+                # Save a copy of the (unprojected) parent solution on the mortar
+                cache.mortars.u_parent[v, i, mortar] = u[v, i_large, j_large, element]
+            end
+            i_large += i_large_step
+            j_large += j_large_step
+        end
+
+        # Interpolate large element face data from buffer to small face locations
+        Trixi.multiply_dimensionwise!(view(cache.mortars.u, 2, :, 1, :, mortar),
+                                      mortar_l2.forward_lower,
+                                      u_buffer)
+        Trixi.multiply_dimensionwise!(view(cache.mortars.u, 2, :, 2, :, mortar),
+                                      mortar_l2.forward_upper,
+                                      u_buffer)
+
+        # After the projection of the constant solution we modify the values
+        # in the first solution variable to no longer be the total water height H = h+b
+        # and instead recover the conservative water height variable `h`.
+        # Basically, unpacking the total water height variable to create the projected local water
+        # height `h` from Eq. 41 in Benov et al.
+        for i in eachnode(dg)
+            cache.mortars.u[2, 1, 1, i, mortar] = max(cache.mortars.u[2, 1, 1, i,
+                                                                      mortar] -
+                                                      cache.mortars.u[2, 4, 1, i,
+                                                                      mortar],
+                                                      equations.threshold_limiter)
+            cache.mortars.u[2, 1, 2, i, mortar] = max(cache.mortars.u[2, 1, 2, i,
+                                                                      mortar] -
+                                                      cache.mortars.u[2, 4, 2, i,
+                                                                      mortar],
+                                                      equations.threshold_limiter)
+
+            # Safety application of velocity desingularization and water height cutoff on the mortars.
+            # Here we use the rather conservative velocity desingularization tolerance of `1e-4`
+            # which works for the AMR testing. The more standard value of `1e-6` works
+            # for well-balancedness testing, but crashes the AMR simulation elixir.
+            #
+            # For details on the motivation of the velocity desingularization see
+            # - A. Chertock, S. Cui, A. Kurganov, T. Wu (2015)
+            #   Well-balanced positivity preserving central-upwind scheme for
+            #   the shallow water system with friction terms
+            #   [DOI: 10.1002/fld.4023](https://doi.org/10.1002/fld.4023)
+
+            # Mortars with the copied or projected solution
+            tol = 1e-4
```

Comment:
```suggestion
            tol = convert(eltype(cache.mortars.u), 1.0e-4)
```

Reply:
Now that we have the additional `threshold_desingularization` available I should be able to grab this tolerance directly. Although this requires some testing.

Reply:
Now I pass this tolerance into the equations constructor as `threshold_desingularization`, so the type conversion should be done under the hood.